Implement single-field property change for multinode edit

RandomShaper opened the pull request:
This allows you to change only one component of any math-typed property during a multi-node edit. The way of telling the editor you want to change only the focused component is to hit __Shift + Enter__. Hitting just __Enter__ will set the whole value, as always.

For instance, you have some sprites at different _x_ coordinates but sharing the same _y_. Now you'll be able to change only the _y_ value by selecting them all, editing the position property, focusing the _y_ field and entering a new value.

This works for vectors, rects, bounding-boxes, transforms, etc.

Closes #7538.

EDIT: This may need testing so now _master_ is unstable we have a great chance.

akien-mga commented on core/math/fieldwise.cpp:
```diff
@@ -0,0 +1,181 @@
+/*************************************************************************/
+/*  a_star.cpp                                                        */
```
Wrong filename :)

akien-mga commented on core/math/fieldwise.h:
```diff
@@ -0,0 +1,40 @@
+/*************************************************************************/
+/*  a_star.h                                                        */
```
Same here.

RandomShaper replied:
Amended and pushed.

bojidar-bg commented on core/math/fieldwise.cpp:
```diff
@@ -0,0 +1,181 @@
+			else if (p_field=="zz") t.basis.elements[2][2]=s.basis.elements[2][2];
+			else if (p_field=="xo") t.origin.x=s.origin.x;
+			else if (p_field=="yo") t.origin.y=s.origin.y;
+			else if (p_field=="zo") t.origin.z=s.origin.z;
```
I think you might be able to use some macro here, not sure about it though.

RandomShaper replied:
Let's macro-ize then!

bojidar-bg commented on core/math/fieldwise.h:
```diff
@@ -0,0 +1,40 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+#ifndef MATH_FIELDWISE_H
+#define MATH_FIELDWISE_H
```
Hmm, why is this file inside `math/`?

RandomShaper replied:
I thought a bit about where to place this. I considered putting it in Variant, but I concluded it would get polluted. Finally, since this only applies to math types I decided to put it in its own file but close to the definitions of the types (at least at the filesystem level).

bojidar-bg replied:
Yeah, but it is still editor-only, so I'm not sure if it should even be in `core`

RandomShaper replied:
Good point. What would be a good place? What I would advise against is coupling this with the property editor stuff because this may serve another purpose in the future.

akien-mga replied:
Note that we also have `core/helper/value_evaluator.h` which is kind of similar (a helper for the inspector) and likely also badly located.

bojidar-bg commented on tools/editor/multi_node_edit.cpp:
```diff
@@ -167,6 +181,11 @@ void MultiNodeEdit::add_node(const NodePath& p_node){
 
+void MultiNodeEdit::set_property_field(const StringName& p_property, const Variant& p_value, const String& p_field) {
+
+	_set_impl(p_property, p_value, p_field);
```
You can probably move all of `_set_impl` here, and then call `set_property_field(p_name, p_value, "")` in `_set`.

RandomShaper replied:
I did it that way to make clear there are two entry points for the same functionality, that is, to keep symmetry, it that makes sense.

bojidar-bg commented on tools/editor/property_editor.cpp:
```diff
@@ -1864,6 +1867,15 @@ void CustomPropertyEditor::_modified(String p_string) {
 
+void CustomPropertyEditor::_emit_changed_whole_or_field() {
+
+	if (!Input::get_singleton()->is_key_pressed(KEY_SHIFT)) {
```
Unconfigurable shortcut (nonblocking). (so, feel free to ignore this)

RandomShaper replied:
Making this configurable would be great, but in my tests with other keys, for instance, Control+Enter was ignored. Anyway, if you can point me on it could be done, I'll try.

bojidar-bg replied:
Well, if it isn't doable, ignore it -- maybe someone else would fix it one day :smiley:

bojidar-bg reviewed (changes requested):
You don't have to do all those changes, feel free to defend if I'm wrong somewhere :wink:

RandomShaper commented on core/math/fieldwise.cpp:
```diff
@@ -0,0 +1,181 @@
+			Vector2 t=p_target;
+
+			     if (p_field=="x") t.x=s.x;
+			else if (p_field=="y") t.y=s.y;
```
Not sure about how to make this good-looking then.

RandomShaper commented:
Now the code looks much prettier, doesn't it? :)

RandomShaper commented:
What about keeping this as it is now and adding a refactor task for both to
the backlog?

El 1 mar. 2017 1:39 p. m., "Rémi Verschelde" <notifications@github.com>
escribió:

> *@akien-mga* commented on this pull request.
> ------------------------------
>
> In core/math/fieldwise.h
> <https://github.com/godotengine/godot/pull/7911#discussion_r103674322>:
>
> > +/* permit persons to whom the Software is furnished to do so, subject to */
> +/* the following conditions:                                             */
> +/*                                                                       */
> +/* The above copyright notice and this permission notice shall be        */
> +/* included in all copies or substantial portions of the Software.       */
> +/*                                                                       */
> +/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
> +/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
> +/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
> +/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
> +/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
> +/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
> +/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
> +/*************************************************************************/
> +#ifndef MATH_FIELDWISE_H
> +#define MATH_FIELDWISE_H
>
> Note that we also have core/helper/value_evaluator.h which is kind of
> similar (a helper for the inspector) and likely also badly located.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/7911#discussion_r103674322>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCtsTh7aTDFS461hp2gHzb3dVqvvLxks5rhWcCgaJpZM4MPJZS>
> .
>

bojidar-bg commented:
@RandomShaper Well, it is prettier indeed, but you are still space-aligning those things...

Maybe we can just move it to `core/helper` as well?

RandomShaper commented:
I need ideas for non-space alignment.

Maybe pad with /**/ ?

El 1 mar. 2017 3:43 p. m., "Bojidar Marinov" <notifications@github.com>
escribió:

> @RandomShaper <https://github.com/RandomShaper> Well, it is prettier
> indeed, but you are still space-aligning those things...
>
> Maybe we can just move it to core/helper as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/7911#issuecomment-283358367>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCtqnNd_4H3u4fuSJlcNpzwXkSqlxNks5rhYQXgaJpZM4MPJZS>
> .
>

akien-mga commented:
I think we can live with it not being aligned, like in all other occurrences where we have big ifs ;)

RandomShaper commented:
If it compiles and works, this is my last push to this PR.

I've moved the files to __core/helper/math_fieldwise.*__, aligned with `/**/` (which looks good) and fixed a huge bug introduced during the macro-ization.

RandomShaper committed:
Implement single-field property change for multinode edit

bojidar-bg reviewed (approved):
Looks good, and works good (for Vector2 at least)

RandomShaper commented:
I've passed Bojidar's OCD test! XD

bojidar-bg commented:
@RandomShaper Well, you could have dropped the else-s and added returns to the ifs........... (no, don't)

RandomShaper commented:
I could but keeping the else out of the macro makes the code reading more natural, IMHO.

Seriously, you need to take a break. :)

akien-mga commented:
See what you've done? https://github.com/akien-mga/godot/commit/0ba67337676ab5131d496af81650ad613eb7906d#diff-c99bc4dcf198a984c35435bc0026dd9dR124

Told you we could live without this alignment ;)

akien-mga commented:
Actually "see what you've done?" for @bojidar-bg too, clang-format sucks with macros:
```diff
-                       TRY_TRANSFER_FIELD("x",normal.x)
-                       else TRY_TRANSFER_FIELD("y",normal.y)
-                       else TRY_TRANSFER_FIELD("z",normal.z)
-                       else TRY_TRANSFER_FIELD("d",d)
+                       TRY_TRANSFER_FIELD("x", normal.x)
+                       else TRY_TRANSFER_FIELD("y", normal.y) else TRY_TRANSFER_FIELD("z", normal.z) else TRY_TRANSFER_FIELD("d", d)
 
-                       return target;
+                                       return target;
```

akien-mga commented:
I'll fix it by disabling clang-format on that code chunk.

RandomShaper commented:
:O

What a mess!

El 5 mar. 2017 4:30 p. m., "Rémi Verschelde" <notifications@github.com>
escribió:

> I'll fix it by disabling clang-format on that code chunk.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/7911#issuecomment-284235531>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCtiMnXDmlgmTPBHhCae10vDwBZ3yYks5ritUogaJpZM4MPJZS>
> .
>

RandomShaper commented:
Maybe those can be formatted with semicolons and curly braces for a better
result.

El 5 mar. 2017 5:03 p. m., "Pedro J. Estébanez" <pedrojrulez@gmail.com>
escribió:

> :O
>
> What a mess!
>
> El 5 mar. 2017 4:30 p. m., "Rémi Verschelde" <notifications@github.com>
> escribió:
>
>> I'll fix it by disabling clang-format on that code chunk.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/7911#issuecomment-284235531>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ALQCtiMnXDmlgmTPBHhCae10vDwBZ3yYks5ritUogaJpZM4MPJZS>
>> .
>>
>

bojidar-bg commented:
Yeah, I guess if we take the semicolon out of the macro, it should work fine..

RandomShaper commented:
I've spent half night setting up my new Git environment to be able to
contribute easily and I'm a bit tired.

So PR welcome :)

El 6 mar. 2017 12:15 p. m., "Bojidar Marinov" <notifications@github.com>
escribió:

> Yeah, I guess if we take the semicolon out of the macro, it should work
> fine..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/7911#issuecomment-284368982>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCtvLmP382MZoMKyhR17QHAOy9lV4Bks5ri-rZgaJpZM4MPJZS>
> .
>
